browser/sodium/sodium.js

```diff
@@ -0,0 +1,158 @@
+const sodium = require('libsodium-wrappers')
+const CryptoProvider = require('../../lib/crypto/cryptoProvider')
+const Platform = require('../platform')
+const { DEFAULT_KDF_ITERATIONS } = require('../../lib/utils/constants')
+
+const encoder = new Platform()
+
+/**
+ * Use PBKDF2 to derive a key of a given length using a specified password
+ * and salt.
+ *
+ * @param {string} password     User-specified password
+ * @param {string} salt         User-specified salt (should be random)
+ * @param {number} length       Length of the key to generate
+ * @param {number} [iterations] Option number of hash iterations to create the seed.
+ *
+ * @returns {Promise<ArrayBuffer>}
+ */
+async function deriveKey(password, salt, length, iterations) {
+  password = maybeToBytes(password)
+  salt = maybeToBytes(salt)
+  const keyMaterial = await window.crypto.subtle.importKey(
+    'raw',
+    password,
+    { name: 'PBKDF2' },
+    false,
+    ['deriveBits', 'deriveKey']
+  )
+  const key = await window.crypto.subtle.deriveKey(
+    {
+      name: 'PBKDF2',
+      salt,
+      iterations,
+      hash: 'SHA-512',
+    },
+    keyMaterial,
+    { name: 'HMAC', hash: { name: 'SHA-1' }, length: length * 8 },
+    true,
+    ['sign', 'verify']
+  )
+  return window.crypto.subtle.exportKey('raw', key)
+}
+
+function maybeToBytes(val) {
```

Comment:
ToBufferOrByteArray?

Reply:
Only a buffer in node. Ensures a UInt8Array - this is an internal compatibility function.